// @bit/primefaces.primeng.paginator/dist/paginator/esm2015/paginator/paginator.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, Component, Input, Output, ChangeDetectorRef, EventEmitter, TemplateRef } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { DropdownModule } from '../dropdown/dropdown';
import { SharedModule } from '../common/shared';
export class Paginator {
    /**
     * @param {?} cd
     */
    constructor(cd) {
        this.cd = cd;
        this.pageLinkSize = 5;
        this.onPageChange = new EventEmitter();
        this.alwaysShow = true;
        this.dropdownScrollHeight = '200px';
        this.currentPageReportTemplate = '{currentPage} of {totalPages}';
        this._totalRecords = 0;
        this._first = 0;
        this._rows = 0;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.updatePaginatorState();
    }
    /**
     * @return {?}
     */
    get totalRecords() {
        return this._totalRecords;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set totalRecords(val) {
        this._totalRecords = val;
        this.updatePageLinks();
        this.updatePaginatorState();
        this.updateFirst();
        this.updateRowsPerPageOptions();
    }
    /**
     * @return {?}
     */
    get first() {
        return this._first;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set first(val) {
        this._first = val;
        this.updatePageLinks();
        this.updatePaginatorState();
    }
    /**
     * @return {?}
     */
    get rows() {
        return this._rows;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set rows(val) {
        this._rows = val;
        this.updatePageLinks();
        this.updatePaginatorState();
    }
    /**
     * @return {?}
     */
    get rowsPerPageOptions() {
        return this._rowsPerPageOptions;
    }
    /**
     * @param {?} val
     * @return {?}
     */
    set rowsPerPageOptions(val) {
        this._rowsPerPageOptions = val;
        this.updateRowsPerPageOptions();
    }
    /**
     * @return {?}
     */
    updateRowsPerPageOptions() {
        if (this.rowsPerPageOptions) {
            this.rowsPerPageItems = [];
            for (let opt of this.rowsPerPageOptions) {
                if (typeof opt == 'object' && opt['showAll']) {
                    this.rowsPerPageItems.push({ label: opt['showAll'], value: this.totalRecords });
                }
                else {
                    this.rowsPerPageItems.push({ label: String(opt), value: opt });
                }
            }
        }
    }
    /**
     * @return {?}
     */
    isFirstPage() {
        return this.getPage() === 0;
    }
    /**
     * @return {?}
     */
    isLastPage() {
        return this.getPage() === this.getPageCount() - 1;
    }
    /**
     * @return {?}
     */
    getPageCount() {
        return Math.ceil(this.totalRecords / this.rows) || 1;
    }
    /**
     * @return {?}
     */
    calculatePageLinkBoundaries() {
        /** @type {?} */
        let numberOfPages = this.getPageCount();
        /** @type {?} */
        let visiblePages = Math.min(this.pageLinkSize, numberOfPages);
        //calculate range, keep current in middle if necessary
        /** @type {?} */
        let start = Math.max(0, Math.ceil(this.getPage() - ((visiblePages) / 2)));
        /** @type {?} */
        let end = Math.min(numberOfPages - 1, start + visiblePages - 1);
        //check when approaching to last page
        /** @type {?} */
        var delta = this.pageLinkSize - (end - start + 1);
        start = Math.max(0, start - delta);
        return [start, end];
    }
    /**
     * @return {?}
     */
    updatePageLinks() {
        this.pageLinks = [];
        /** @type {?} */
        let boundaries = this.calculatePageLinkBoundaries();
        /** @type {?} */
        let start = boundaries[0];
        /** @type {?} */
        let end = boundaries[1];
        for (let i = start; i <= end; i++) {
            this.pageLinks.push(i + 1);
        }
    }
    /**
     * @param {?} p
     * @return {?}
     */
    changePage(p) {
        /** @type {?} */
        var pc = this.getPageCount();
        if (p >= 0 && p < pc) {
            this.first = this.rows * p;
            /** @type {?} */
            var state = {
                page: p,
                first: this.first,
                rows: this.rows,
                pageCount: pc
            };
            this.updatePageLinks();
            this.onPageChange.emit(state);
            this.updatePaginatorState();
        }
    }
    /**
     * @return {?}
     */
    updateFirst() {
        /** @type {?} */
        const page = this.getPage();
        if (page > 0 && (this.first >= this.totalRecords)) {
            Promise.resolve(null).then((/**
             * @return {?}
             */
            () => this.changePage(page - 1)));
        }
    }
    /**
     * @return {?}
     */
    getPage() {
        return Math.floor(this.first / this.rows);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    changePageToFirst(event) {
        if (!this.isFirstPage()) {
            this.changePage(0);
        }
        event.preventDefault();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    changePageToPrev(event) {
        this.changePage(this.getPage() - 1);
        event.preventDefault();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    changePageToNext(event) {
        this.changePage(this.getPage() + 1);
        event.preventDefault();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    changePageToLast(event) {
        if (!this.isLastPage()) {
            this.changePage(this.getPageCount() - 1);
        }
        event.preventDefault();
    }
    /**
     * @param {?} event
     * @param {?} page
     * @return {?}
     */
    onPageLinkClick(event, page) {
        this.changePage(page);
        event.preventDefault();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onRppChange(event) {
        this.changePage(this.getPage());
    }
    /**
     * @return {?}
     */
    updatePaginatorState() {
        this.paginatorState = {
            page: this.getPage(),
            pageCount: this.getPageCount(),
            rows: this.rows,
            first: this.first,
            totalRecords: this.totalRecords
        };
    }
    /**
     * @return {?}
     */
    get currentPageReport() {
        return this.currentPageReportTemplate
            .replace("{currentPage}", (this.getPage() + 1).toString())
            .replace("{totalPages}", this.getPageCount().toString());
    }
}
Paginator.decorators = [
    { type: Component, args: [{
                selector: 'p-paginator',
                template: `
        <div [class]="styleClass" [ngStyle]="style" [ngClass]="'ui-paginator ui-widget ui-widget-header ui-unselectable-text ui-helper-clearfix'"
            *ngIf="alwaysShow ? true : (pageLinks && pageLinks.length > 1)">
            <div class="ui-paginator-left-content" *ngIf="templateLeft">
                <ng-container *ngTemplateOutlet="templateLeft; context: {$implicit: paginatorState}"></ng-container>
            </div>
            <span class="ui-paginator-current" *ngIf="showCurrentPageReport">{{currentPageReport}}</span>
            <a [attr.tabindex]="isFirstPage() ? null : '0'" class="ui-paginator-first ui-paginator-element ui-state-default ui-corner-all"
                    (click)="changePageToFirst($event)" (keydown.enter)="changePageToFirst($event)" [ngClass]="{'ui-state-disabled':isFirstPage()}" [tabindex]="isFirstPage() ? -1 : null">
                <span class="ui-paginator-icon pi pi-step-backward"></span>
            </a>
            <a tabindex="0" [attr.tabindex]="isFirstPage() ? null : '0'" class="ui-paginator-prev ui-paginator-element ui-state-default ui-corner-all"
                    (click)="changePageToPrev($event)" (keydown.enter)="changePageToPrev($event)" [ngClass]="{'ui-state-disabled':isFirstPage()}" [tabindex]="isFirstPage() ? -1 : null">
                <span class="ui-paginator-icon pi pi-caret-left"></span>
            </a>
            <span class="ui-paginator-pages">
                <a tabindex="0" *ngFor="let pageLink of pageLinks" class="ui-paginator-page ui-paginator-element ui-state-default ui-corner-all"
                    (click)="onPageLinkClick($event, pageLink - 1)" (keydown.enter)="onPageLinkClick($event, pageLink - 1)" [ngClass]="{'ui-state-active': (pageLink-1 == getPage())}">{{pageLink}}</a>
            </span>
            <a [attr.tabindex]="isLastPage() ? null : '0'" class="ui-paginator-next ui-paginator-element ui-state-default ui-corner-all"
                    (click)="changePageToNext($event)" (keydown.enter)="changePageToNext($event)" [ngClass]="{'ui-state-disabled':isLastPage()}" [tabindex]="isLastPage() ? -1 : null">
                <span class="ui-paginator-icon pi pi-caret-right"></span>
            </a>
            <a [attr.tabindex]="isLastPage() ? null : '0'" class="ui-paginator-last ui-paginator-element ui-state-default ui-corner-all"
                    (click)="changePageToLast($event)" (keydown.enter)="changePageToLast($event)" [ngClass]="{'ui-state-disabled':isLastPage()}" [tabindex]="isLastPage() ? -1 : null">
                <span class="ui-paginator-icon pi pi-step-forward"></span>
            </a>
            <p-dropdown [options]="rowsPerPageItems" [(ngModel)]="rows" *ngIf="rowsPerPageOptions" 
                (onChange)="onRppChange($event)" [appendTo]="dropdownAppendTo" [scrollHeight]="dropdownScrollHeight"></p-dropdown>
            <div class="ui-paginator-right-content" *ngIf="templateRight">
                <ng-container *ngTemplateOutlet="templateRight; context: {$implicit: paginatorState}"></ng-container>
            </div>
        </div>
    `
            }] }
];
/** @nocollapse */
Paginator.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
Paginator.propDecorators = {
    pageLinkSize: [{ type: Input }],
    onPageChange: [{ type: Output }],
    style: [{ type: Input }],
    styleClass: [{ type: Input }],
    alwaysShow: [{ type: Input }],
    templateLeft: [{ type: Input }],
    templateRight: [{ type: Input }],
    dropdownAppendTo: [{ type: Input }],
    dropdownScrollHeight: [{ type: Input }],
    currentPageReportTemplate: [{ type: Input }],
    showCurrentPageReport: [{ type: Input }],
    totalRecords: [{ type: Input }],
    first: [{ type: Input }],
    rows: [{ type: Input }],
    rowsPerPageOptions: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    Paginator.prototype.pageLinkSize;
    /** @type {?} */
    Paginator.prototype.onPageChange;
    /** @type {?} */
    Paginator.prototype.style;
    /** @type {?} */
    Paginator.prototype.styleClass;
    /** @type {?} */
    Paginator.prototype.alwaysShow;
    /** @type {?} */
    Paginator.prototype.templateLeft;
    /** @type {?} */
    Paginator.prototype.templateRight;
    /** @type {?} */
    Paginator.prototype.dropdownAppendTo;
    /** @type {?} */
    Paginator.prototype.dropdownScrollHeight;
    /** @type {?} */
    Paginator.prototype.currentPageReportTemplate;
    /** @type {?} */
    Paginator.prototype.showCurrentPageReport;
    /** @type {?} */
    Paginator.prototype.pageLinks;
    /** @type {?} */
    Paginator.prototype._totalRecords;
    /** @type {?} */
    Paginator.prototype._first;
    /** @type {?} */
    Paginator.prototype._rows;
    /** @type {?} */
    Paginator.prototype._rowsPerPageOptions;
    /** @type {?} */
    Paginator.prototype.rowsPerPageItems;
    /** @type {?} */
    Paginator.prototype.paginatorState;
    /**
     * @type {?}
     * @private
     */
    Paginator.prototype.cd;
}
export class PaginatorModule {
}
PaginatorModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, DropdownModule, FormsModule, SharedModule],
                exports: [Paginator, DropdownModule, FormsModule, SharedModule],
                declarations: [Paginator]
            },] }
];
