// @bit/primefaces.primeng.dropdown/dist/dropdown/esm5/dropdown/dropdown.js
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ScrollingModule, CdkVirtualScrollViewport } from '@angular/cdk/scrolling';
import { NgModule, Component, ElementRef, Input, Output, Renderer2, EventEmitter, ContentChildren, QueryList, ViewChild, TemplateRef, forwardRef, ChangeDetectorRef, NgZone } from '@angular/core';
import { trigger, state, style, transition, animate } from '@angular/animations';
import { CommonModule } from '@angular/common';
import { SharedModule, PrimeTemplate } from '../common/shared';
import { DomHandler } from '../dom/domhandler';
import { ObjectUtils } from '../utils/objectutils';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
/** @type {?} */
export var DROPDOWN_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef((/**
     * @return {?}
     */
    function () { return Dropdown; })),
    multi: true
};
var DropdownItem = /** @class */ (function () {
    function DropdownItem() {
        this.onClick = new EventEmitter();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    DropdownItem.prototype.onOptionClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.onClick.emit({
            originalEvent: event,
            option: this.option
        });
    };
    DropdownItem.decorators = [
        { type: Component, args: [{
                    selector: 'p-dropdownItem',
                    template: "\n        <li (click)=\"onOptionClick($event)\" role=\"option\"\n            [attr.aria-label]=\"option.label\"\n            [ngStyle]=\"{'height': itemSize + 'px'}\"\n            [ngClass]=\"{'ui-dropdown-item ui-corner-all':true,\n                                                'ui-state-highlight': selected,\n                                                'ui-state-disabled':(option.disabled),\n                                                'ui-dropdown-item-empty': !option.label||option.label.length === 0}\">\n            <span *ngIf=\"!template\">{{option.label||'empty'}}</span>\n            <ng-container *ngTemplateOutlet=\"template; context: {$implicit: option}\"></ng-container>\n        </li>\n    "
                }] }
    ];
    DropdownItem.propDecorators = {
        option: [{ type: Input }],
        selected: [{ type: Input }],
        disabled: [{ type: Input }],
        visible: [{ type: Input }],
        itemSize: [{ type: Input }],
        template: [{ type: Input }],
        onClick: [{ type: Output }]
    };
    return DropdownItem;
}());
export { DropdownItem };
if (false) {
    /** @type {?} */
    DropdownItem.prototype.option;
    /** @type {?} */
    DropdownItem.prototype.selected;
    /** @type {?} */
    DropdownItem.prototype.disabled;
    /** @type {?} */
    DropdownItem.prototype.visible;
    /** @type {?} */
    DropdownItem.prototype.itemSize;
    /** @type {?} */
    DropdownItem.prototype.template;
    /** @type {?} */
    DropdownItem.prototype.onClick;
}
var Dropdown = /** @class */ (function () {
    function Dropdown(el, renderer, cd, zone) {
        this.el = el;
        this.renderer = renderer;
        this.cd = cd;
        this.zone = zone;
        this.scrollHeight = '200px';
        this.filterBy = 'label';
        this.resetFilterOnHide = false;
        this.dropdownIcon = 'pi pi-chevron-down';
        this.autoDisplayFirst = true;
        this.emptyFilterMessage = 'No results found';
        this.autoZIndex = true;
        this.baseZIndex = 0;
        this.showTransitionOptions = '225ms ease-out';
        this.hideTransitionOptions = '195ms ease-in';
        this.onChange = new EventEmitter();
        this.onFocus = new EventEmitter();
        this.onBlur = new EventEmitter();
        this.onClick = new EventEmitter();
        this.onShow = new EventEmitter();
        this.onHide = new EventEmitter();
        this.onModelChange = (/**
         * @return {?}
         */
        function () { });
        this.onModelTouched = (/**
         * @return {?}
         */
        function () { });
    }
    Object.defineProperty(Dropdown.prototype, "autoWidth", {
        get: /**
         * @return {?}
         */
        function () {
            return this._autoWidth;
        },
        set: /**
         * @param {?} _autoWidth
         * @return {?}
         */
        function (_autoWidth) {
            this._autoWidth = _autoWidth;
            console.log("Setting autoWidth has no effect as automatic width calculation is removed for better perfomance.");
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Dropdown.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        set: /**
         * @param {?} _disabled
         * @return {?}
         */
        function (_disabled) {
            if (_disabled)
                this.focused = false;
            this._disabled = _disabled;
            this.cd.detectChanges();
        },
        enumerable: true,
        configurable: true
    });
    ;
    /**
     * @return {?}
     */
    Dropdown.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.templates.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            switch (item.getType()) {
                case 'item':
                    _this.itemTemplate = item.template;
                    break;
                case 'selectedItem':
                    _this.selectedItemTemplate = item.template;
                    break;
                case 'group':
                    _this.groupTemplate = item.template;
                    break;
                default:
                    _this.itemTemplate = item.template;
                    break;
            }
        }));
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.optionsToDisplay = this.options;
        this.updateSelectedOption(null);
    };
    Object.defineProperty(Dropdown.prototype, "options", {
        get: /**
         * @return {?}
         */
        function () {
            return this._options;
        },
        set: /**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            /** @type {?} */
            var opts = this.optionLabel ? ObjectUtils.generateSelectItems(val, this.optionLabel) : val;
            this._options = opts;
            this.optionsToDisplay = this._options;
            this.updateSelectedOption(this.value);
            this.optionsChanged = true;
            if (this.filterValue && this.filterValue.length) {
                this.activateFilter();
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    Dropdown.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.editable) {
            this.updateEditableLabel();
        }
    };
    Object.defineProperty(Dropdown.prototype, "label", {
        get: /**
         * @return {?}
         */
        function () {
            return (this.selectedOption ? this.selectedOption.label : null);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    Dropdown.prototype.updateEditableLabel = /**
     * @return {?}
     */
    function () {
        if (this.editableInputViewChild && this.editableInputViewChild.nativeElement) {
            this.editableInputViewChild.nativeElement.value = (this.selectedOption ? this.selectedOption.label : this.value || '');
        }
    };
    /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    Dropdown.prototype.onItemClick = /**
     * @param {?} event
     * @param {?} index
     * @return {?}
     */
    function (event, index) {
        var _this = this;
        /** @type {?} */
        var option = event.option;
        this.itemClick = true;
        if (this.virtualScroll) {
            this.virtualScrollSelectedIndex = index;
        }
        if (!option.disabled) {
            this.selectItem(event, option);
            this.focusViewChild.nativeElement.focus();
            this.filled = true;
        }
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.hide();
        }), 150);
    };
    /**
     * @param {?} event
     * @param {?} option
     * @return {?}
     */
    Dropdown.prototype.selectItem = /**
     * @param {?} event
     * @param {?} option
     * @return {?}
     */
    function (event, option) {
        if (this.selectedOption != option) {
            this.selectedOption = option;
            this.value = option.value;
            this.onModelChange(this.value);
            this.updateEditableLabel();
            this.onChange.emit({
                originalEvent: event.originalEvent,
                value: this.value
            });
        }
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.ngAfterViewChecked = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.optionsChanged && this.overlayVisible) {
            this.optionsChanged = false;
            this.zone.runOutsideAngular((/**
             * @return {?}
             */
            function () {
                setTimeout((/**
                 * @return {?}
                 */
                function () {
                    _this.alignOverlay();
                }), 1);
            }));
        }
        if (this.selectedOptionUpdated && this.itemsWrapper) {
            /** @type {?} */
            var selectedItem = DomHandler.findSingle(this.overlay, 'li.ui-state-highlight');
            if (selectedItem) {
                DomHandler.scrollInView(this.itemsWrapper, DomHandler.findSingle(this.overlay, 'li.ui-state-highlight'));
            }
            this.selectedOptionUpdated = false;
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    Dropdown.prototype.writeValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (this.filter) {
            this.resetFilter();
        }
        this.value = value;
        this.updateSelectedOption(value);
        this.updateEditableLabel();
        this.updateFilledState();
        this.cd.markForCheck();
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.resetFilter = /**
     * @return {?}
     */
    function () {
        if (this.filterViewChild && this.filterViewChild.nativeElement) {
            this.filterValue = null;
            this.filterViewChild.nativeElement.value = '';
        }
        this.optionsToDisplay = this.options;
    };
    /**
     * @param {?} val
     * @return {?}
     */
    Dropdown.prototype.updateSelectedOption = /**
     * @param {?} val
     * @return {?}
     */
    function (val) {
        this.selectedOption = this.findOption(val, this.optionsToDisplay);
        if (this.autoDisplayFirst && !this.placeholder && !this.selectedOption && this.optionsToDisplay && this.optionsToDisplay.length && !this.editable) {
            this.selectedOption = this.optionsToDisplay[0];
        }
        this.selectedOptionUpdated = true;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    Dropdown.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onModelChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    Dropdown.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onModelTouched = fn;
    };
    /**
     * @param {?} val
     * @return {?}
     */
    Dropdown.prototype.setDisabledState = /**
     * @param {?} val
     * @return {?}
     */
    function (val) {
        this.disabled = val;
    };
    /**
     * @param {?} event
     * @return {?}
     */
    Dropdown.prototype.onMouseclick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.disabled || this.readonly) {
            return;
        }
        this.onClick.emit(event);
        this.selfClick = true;
        this.clearClick = DomHandler.hasClass(event.target, 'ui-dropdown-clear-icon');
        if (!this.itemClick && !this.clearClick) {
            this.focusViewChild.nativeElement.focus();
            if (this.overlayVisible)
                this.hide();
            else
                this.show();
            this.cd.detectChanges();
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    Dropdown.prototype.onEditableInputClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.itemClick = true;
        this.bindDocumentClickListener();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    Dropdown.prototype.onEditableInputFocus = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.focused = true;
        this.hide();
        this.onFocus.emit(event);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    Dropdown.prototype.onEditableInputChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.value = event.target.value;
        this.updateSelectedOption(this.value);
        this.onModelChange(this.value);
        this.onChange.emit({
            originalEvent: event,
            value: this.value
        });
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.show = /**
     * @return {?}
     */
    function () {
        this.overlayVisible = true;
    };
    /**
     * @param {?} event
     * @return {?}
     */
    Dropdown.prototype.onOverlayAnimationStart = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        switch (event.toState) {
            case 'visible':
                this.overlay = event.element;
                this.itemsWrapper = DomHandler.findSingle(this.overlay, '.ui-dropdown-items-wrapper');
                this.appendOverlay();
                if (this.autoZIndex) {
                    this.overlay.style.zIndex = String(this.baseZIndex + (++DomHandler.zindex));
                }
                this.alignOverlay();
                this.bindDocumentClickListener();
                this.bindDocumentResizeListener();
                if (this.options && this.options.length) {
                    if (!this.virtualScroll) {
                        /** @type {?} */
                        var selectedListItem = DomHandler.findSingle(this.itemsWrapper, '.ui-dropdown-item.ui-state-highlight');
                        if (selectedListItem) {
                            DomHandler.scrollInView(this.itemsWrapper, selectedListItem);
                        }
                    }
                }
                if (this.filterViewChild && this.filterViewChild.nativeElement) {
                    this.filterViewChild.nativeElement.focus();
                }
                this.onShow.emit(event);
                break;
            case 'void':
                this.onOverlayHide();
                break;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    Dropdown.prototype.scrollToSelectedVirtualScrollElement = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        if (!this.virtualAutoScrolled) {
            if (this.filter && !this.resetFilterOnHide) {
                /** @type {?} */
                var index = this.optionsToDisplay.findIndex((/**
                 * @param {?} option
                 * @return {?}
                 */
                function (option) { return option.value === _this.value; }));
                if (event == 0 && index > 0) {
                    this.viewPort.scrollToIndex(index, 'auto');
                }
            }
            else {
                if (event == 0 && this.virtualScrollSelectedIndex > 0) {
                    this.viewPort.scrollToIndex(this.virtualScrollSelectedIndex, 'auto');
                }
            }
            this.virtualAutoScrolled = true;
        }
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.appendOverlay = /**
     * @return {?}
     */
    function () {
        if (this.appendTo) {
            if (this.appendTo === 'body')
                document.body.appendChild(this.overlay);
            else
                DomHandler.appendChild(this.overlay, this.appendTo);
            this.overlay.style.minWidth = DomHandler.getWidth(this.containerViewChild.nativeElement) + 'px';
        }
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.restoreOverlayAppend = /**
     * @return {?}
     */
    function () {
        if (this.overlay && this.appendTo) {
            this.el.nativeElement.appendChild(this.overlay);
        }
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.hide = /**
     * @return {?}
     */
    function () {
        this.overlayVisible = false;
        if (this.filter && this.resetFilterOnHide) {
            this.resetFilter();
        }
        if (this.virtualScroll) {
            this.virtualAutoScrolled = false;
        }
        this.cd.markForCheck();
        this.onHide.emit(event);
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.alignOverlay = /**
     * @return {?}
     */
    function () {
        if (this.overlay) {
            if (this.appendTo)
                DomHandler.absolutePosition(this.overlay, this.containerViewChild.nativeElement);
            else
                DomHandler.relativePosition(this.overlay, this.containerViewChild.nativeElement);
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    Dropdown.prototype.onInputFocus = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.focused = true;
        this.onFocus.emit(event);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    Dropdown.prototype.onInputBlur = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.focused = false;
        this.onModelTouched();
        this.onBlur.emit(event);
    };
    /**
     * @param {?} index
     * @return {?}
     */
    Dropdown.prototype.findPrevEnabledOption = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        /** @type {?} */
        var prevEnabledOption;
        if (this.optionsToDisplay && this.optionsToDisplay.length) {
            for (var i = (index - 1); 0 <= i; i--) {
                /** @type {?} */
                var option = this.optionsToDisplay[i];
                if (option.disabled) {
                    continue;
                }
                else {
                    prevEnabledOption = option;
                    break;
                }
            }
            if (!prevEnabledOption) {
                for (var i = this.optionsToDisplay.length - 1; i >= index; i--) {
                    /** @type {?} */
                    var option = this.optionsToDisplay[i];
                    if (option.disabled) {
                        continue;
                    }
                    else {
                        prevEnabledOption = option;
                        break;
                    }
                }
            }
        }
        return prevEnabledOption;
    };
    /**
     * @param {?} index
     * @return {?}
     */
    Dropdown.prototype.findNextEnabledOption = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        /** @type {?} */
        var nextEnabledOption;
        if (this.optionsToDisplay && this.optionsToDisplay.length) {
            for (var i = (index + 1); index < (this.optionsToDisplay.length - 1); i++) {
                /** @type {?} */
                var option = this.optionsToDisplay[i];
                if (option.disabled) {
                    continue;
                }
                else {
                    nextEnabledOption = option;
                    break;
                }
            }
            if (!nextEnabledOption) {
                for (var i = 0; i < index; i++) {
                    /** @type {?} */
                    var option = this.optionsToDisplay[i];
                    if (option.disabled) {
                        continue;
                    }
                    else {
                        nextEnabledOption = option;
                        break;
                    }
                }
            }
        }
        return nextEnabledOption;
    };
    /**
     * @param {?} event
     * @param {?} search
     * @return {?}
     */
    Dropdown.prototype.onKeydown = /**
     * @param {?} event
     * @param {?} search
     * @return {?}
     */
    function (event, search) {
        if (this.readonly || !this.optionsToDisplay || this.optionsToDisplay.length === null) {
            return;
        }
        switch (event.which) {
            //down
            case 40:
                if (!this.overlayVisible && event.altKey) {
                    this.show();
                }
                else {
                    if (this.group) {
                        /** @type {?} */
                        var selectedItemIndex = this.selectedOption ? this.findOptionGroupIndex(this.selectedOption.value, this.optionsToDisplay) : -1;
                        if (selectedItemIndex !== -1) {
                            /** @type {?} */
                            var nextItemIndex = selectedItemIndex.itemIndex + 1;
                            if (nextItemIndex < (this.optionsToDisplay[selectedItemIndex.groupIndex].items.length)) {
                                this.selectItem(event, this.optionsToDisplay[selectedItemIndex.groupIndex].items[nextItemIndex]);
                                this.selectedOptionUpdated = true;
                            }
                            else if (this.optionsToDisplay[selectedItemIndex.groupIndex + 1]) {
                                this.selectItem(event, this.optionsToDisplay[selectedItemIndex.groupIndex + 1].items[0]);
                                this.selectedOptionUpdated = true;
                            }
                        }
                        else {
                            this.selectItem(event, this.optionsToDisplay[0].items[0]);
                        }
                    }
                    else {
                        /** @type {?} */
                        var selectedItemIndex = this.selectedOption ? this.findOptionIndex(this.selectedOption.value, this.optionsToDisplay) : -1;
                        /** @type {?} */
                        var nextEnabledOption = this.findNextEnabledOption(selectedItemIndex);
                        if (nextEnabledOption) {
                            this.selectItem(event, nextEnabledOption);
                            this.selectedOptionUpdated = true;
                        }
                    }
                }
                event.preventDefault();
                break;
            //up
            case 38:
                if (this.group) {
                    /** @type {?} */
                    var selectedItemIndex = this.selectedOption ? this.findOptionGroupIndex(this.selectedOption.value, this.optionsToDisplay) : -1;
                    if (selectedItemIndex !== -1) {
                        /** @type {?} */
                        var prevItemIndex = selectedItemIndex.itemIndex - 1;
                        if (prevItemIndex >= 0) {
                            this.selectItem(event, this.optionsToDisplay[selectedItemIndex.groupIndex].items[prevItemIndex]);
                            this.selectedOptionUpdated = true;
                        }
                        else if (prevItemIndex < 0) {
                            /** @type {?} */
                            var prevGroup = this.optionsToDisplay[selectedItemIndex.groupIndex - 1];
                            if (prevGroup) {
                                this.selectItem(event, prevGroup.items[prevGroup.items.length - 1]);
                                this.selectedOptionUpdated = true;
                            }
                        }
                    }
                }
                else {
                    /** @type {?} */
                    var selectedItemIndex = this.selectedOption ? this.findOptionIndex(this.selectedOption.value, this.optionsToDisplay) : -1;
                    /** @type {?} */
                    var prevEnabledOption = this.findPrevEnabledOption(selectedItemIndex);
                    if (prevEnabledOption) {
                        this.selectItem(event, prevEnabledOption);
                        this.selectedOptionUpdated = true;
                    }
                }
                event.preventDefault();
                break;
            //space
            case 32:
            case 32:
                if (!this.overlayVisible) {
                    this.show();
                    event.preventDefault();
                }
                break;
            //enter
            case 13:
                if (!this.filter || (this.optionsToDisplay && this.optionsToDisplay.length > 0)) {
                    this.hide();
                }
                event.preventDefault();
                break;
            //escape and tab
            case 27:
            case 9:
                this.hide();
                break;
            //search item based on keyboard input
            default:
                if (search) {
                    this.search(event);
                }
                break;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    Dropdown.prototype.search = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        if (this.searchTimeout) {
            clearTimeout(this.searchTimeout);
        }
        /** @type {?} */
        var char = String.fromCharCode(event.keyCode);
        this.previousSearchChar = this.currentSearchChar;
        this.currentSearchChar = char;
        if (this.previousSearchChar === this.currentSearchChar)
            this.searchValue = this.currentSearchChar;
        else
            this.searchValue = this.searchValue ? this.searchValue + char : char;
        /** @type {?} */
        var newOption;
        if (this.group) {
            /** @type {?} */
            var searchIndex = this.selectedOption ? this.findOptionGroupIndex(this.selectedOption.value, this.optionsToDisplay) : { groupIndex: 0, itemIndex: 0 };
            newOption = this.searchOptionWithinGroup(searchIndex);
        }
        else {
            /** @type {?} */
            var searchIndex = this.selectedOption ? this.findOptionIndex(this.selectedOption.value, this.optionsToDisplay) : -1;
            newOption = this.searchOption(++searchIndex);
        }
        if (newOption) {
            this.selectItem(event, newOption);
            this.selectedOptionUpdated = true;
        }
        this.searchTimeout = setTimeout((/**
         * @return {?}
         */
        function () {
            _this.searchValue = null;
        }), 250);
    };
    /**
     * @param {?} index
     * @return {?}
     */
    Dropdown.prototype.searchOption = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        /** @type {?} */
        var option;
        if (this.searchValue) {
            option = this.searchOptionInRange(index, this.optionsToDisplay.length);
            if (!option) {
                option = this.searchOptionInRange(0, index);
            }
        }
        return option;
    };
    /**
     * @param {?} start
     * @param {?} end
     * @return {?}
     */
    Dropdown.prototype.searchOptionInRange = /**
     * @param {?} start
     * @param {?} end
     * @return {?}
     */
    function (start, end) {
        for (var i = start; i < end; i++) {
            /** @type {?} */
            var opt = this.optionsToDisplay[i];
            if (opt.label.toLowerCase().startsWith(this.searchValue.toLowerCase())) {
                return opt;
            }
        }
        return null;
    };
    /**
     * @param {?} index
     * @return {?}
     */
    Dropdown.prototype.searchOptionWithinGroup = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        /** @type {?} */
        var option;
        if (this.searchValue) {
            for (var i = index.groupIndex; i < this.optionsToDisplay.length; i++) {
                for (var j = (index.groupIndex === i) ? (index.itemIndex + 1) : 0; j < this.optionsToDisplay[i].items.length; j++) {
                    /** @type {?} */
                    var opt = this.optionsToDisplay[i].items[j];
                    if (opt.label.toLowerCase().startsWith(this.searchValue.toLowerCase())) {
                        return opt;
                    }
                }
            }
            if (!option) {
                for (var i = 0; i <= index.groupIndex; i++) {
                    for (var j = 0; j < ((index.groupIndex === i) ? index.itemIndex : this.optionsToDisplay[i].items.length); j++) {
                        /** @type {?} */
                        var opt = this.optionsToDisplay[i].items[j];
                        if (opt.label.toLowerCase().startsWith(this.searchValue.toLowerCase())) {
                            return opt;
                        }
                    }
                }
            }
        }
        return null;
    };
    /**
     * @param {?} val
     * @param {?} opts
     * @return {?}
     */
    Dropdown.prototype.findOptionIndex = /**
     * @param {?} val
     * @param {?} opts
     * @return {?}
     */
    function (val, opts) {
        /** @type {?} */
        var index = -1;
        if (opts) {
            for (var i = 0; i < opts.length; i++) {
                if ((val == null && opts[i].value == null) || ObjectUtils.equals(val, opts[i].value, this.dataKey)) {
                    index = i;
                    break;
                }
            }
        }
        return index;
    };
    /**
     * @param {?} val
     * @param {?} opts
     * @return {?}
     */
    Dropdown.prototype.findOptionGroupIndex = /**
     * @param {?} val
     * @param {?} opts
     * @return {?}
     */
    function (val, opts) {
        /** @type {?} */
        var groupIndex;
        /** @type {?} */
        var itemIndex;
        if (opts) {
            for (var i = 0; i < opts.length; i++) {
                groupIndex = i;
                itemIndex = this.findOptionIndex(val, opts[i].items);
                if (itemIndex !== -1) {
                    break;
                }
            }
        }
        if (itemIndex !== -1) {
            return { groupIndex: groupIndex, itemIndex: itemIndex };
        }
        else {
            return -1;
        }
    };
    /**
     * @param {?} val
     * @param {?} opts
     * @param {?=} inGroup
     * @return {?}
     */
    Dropdown.prototype.findOption = /**
     * @param {?} val
     * @param {?} opts
     * @param {?=} inGroup
     * @return {?}
     */
    function (val, opts, inGroup) {
        var e_1, _a;
        if (this.group && !inGroup) {
            /** @type {?} */
            var opt = void 0;
            if (opts && opts.length) {
                try {
                    for (var opts_1 = __values(opts), opts_1_1 = opts_1.next(); !opts_1_1.done; opts_1_1 = opts_1.next()) {
                        var optgroup = opts_1_1.value;
                        opt = this.findOption(val, optgroup.items, true);
                        if (opt) {
                            break;
                        }
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (opts_1_1 && !opts_1_1.done && (_a = opts_1.return)) _a.call(opts_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
            return opt;
        }
        else {
            /** @type {?} */
            var index = this.findOptionIndex(val, opts);
            return (index != -1) ? opts[index] : null;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    Dropdown.prototype.onFilter = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var inputValue = event.target.value;
        if (inputValue && inputValue.length) {
            this.filterValue = inputValue;
            this.activateFilter();
        }
        else {
            this.filterValue = null;
            this.optionsToDisplay = this.options;
        }
        this.optionsChanged = true;
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.activateFilter = /**
     * @return {?}
     */
    function () {
        var e_2, _a;
        /** @type {?} */
        var searchFields = this.filterBy.split(',');
        if (this.options && this.options.length) {
            if (this.group) {
                /** @type {?} */
                var filteredGroups = [];
                try {
                    for (var _b = __values(this.options), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var optgroup = _c.value;
                        /** @type {?} */
                        var filteredSubOptions = ObjectUtils.filter(optgroup.items, searchFields, this.filterValue);
                        if (filteredSubOptions && filteredSubOptions.length) {
                            filteredGroups.push({
                                label: optgroup.label,
                                value: optgroup.value,
                                items: filteredSubOptions
                            });
                        }
                    }
                }
                catch (e_2_1) { e_2 = { error: e_2_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_2) throw e_2.error; }
                }
                this.optionsToDisplay = filteredGroups;
            }
            else {
                this.optionsToDisplay = ObjectUtils.filter(this.options, searchFields, this.filterValue);
            }
            this.optionsChanged = true;
        }
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.applyFocus = /**
     * @return {?}
     */
    function () {
        if (this.editable)
            DomHandler.findSingle(this.el.nativeElement, '.ui-dropdown-label.ui-inputtext').focus();
        else
            DomHandler.findSingle(this.el.nativeElement, 'input[readonly]').focus();
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.focus = /**
     * @return {?}
     */
    function () {
        this.applyFocus();
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.bindDocumentClickListener = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (!this.documentClickListener) {
            this.documentClickListener = this.renderer.listen('document', 'click', (/**
             * @return {?}
             */
            function () {
                if (!_this.selfClick && !_this.itemClick) {
                    _this.hide();
                    _this.unbindDocumentClickListener();
                }
                _this.clearClickState();
                _this.cd.markForCheck();
            }));
        }
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.clearClickState = /**
     * @return {?}
     */
    function () {
        this.selfClick = false;
        this.itemClick = false;
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.unbindDocumentClickListener = /**
     * @return {?}
     */
    function () {
        if (this.documentClickListener) {
            this.documentClickListener();
            this.documentClickListener = null;
        }
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.bindDocumentResizeListener = /**
     * @return {?}
     */
    function () {
        this.documentResizeListener = this.onWindowResize.bind(this);
        window.addEventListener('resize', this.documentResizeListener);
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.unbindDocumentResizeListener = /**
     * @return {?}
     */
    function () {
        if (this.documentResizeListener) {
            window.removeEventListener('resize', this.documentResizeListener);
            this.documentResizeListener = null;
        }
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.onWindowResize = /**
     * @return {?}
     */
    function () {
        if (!DomHandler.isAndroid()) {
            this.hide();
        }
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.updateFilledState = /**
     * @return {?}
     */
    function () {
        this.filled = (this.selectedOption != null);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    Dropdown.prototype.clear = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.clearClick = true;
        this.value = null;
        this.onModelChange(this.value);
        this.onChange.emit({
            originalEvent: event,
            value: this.value
        });
        this.updateSelectedOption(this.value);
        this.updateEditableLabel();
        this.updateFilledState();
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.onOverlayHide = /**
     * @return {?}
     */
    function () {
        this.unbindDocumentClickListener();
        this.unbindDocumentResizeListener();
        this.overlay = null;
        this.itemsWrapper = null;
    };
    /**
     * @return {?}
     */
    Dropdown.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.restoreOverlayAppend();
        this.onOverlayHide();
    };
    Dropdown.decorators = [
        { type: Component, args: [{
                    selector: 'p-dropdown',
                    template: "\n         <div #container [ngClass]=\"{'ui-dropdown ui-widget ui-state-default ui-corner-all ui-helper-clearfix':true,\n            'ui-state-disabled':disabled, 'ui-dropdown-open':overlayVisible, 'ui-state-focus':focused, 'ui-dropdown-clearable': showClear && !disabled}\"\n            (click)=\"onMouseclick($event)\" [ngStyle]=\"style\" [class]=\"styleClass\">\n            <div class=\"ui-helper-hidden-accessible\">\n                <input #in [attr.id]=\"inputId\" type=\"text\" [attr.aria-label]=\"selectedOption ? selectedOption.label : ' '\" readonly (focus)=\"onInputFocus($event)\" aria-haspopup=\"listbox\"\n                    (blur)=\"onInputBlur($event)\" (keydown)=\"onKeydown($event, true)\" [disabled]=\"disabled\" [attr.tabindex]=\"tabindex\" [attr.autofocus]=\"autofocus\">\n            </div>\n            <div class=\"ui-helper-hidden-accessible ui-dropdown-hidden-select\">\n                <select [attr.required]=\"required\" [attr.name]=\"name\" tabindex=\"-1\" aria-hidden=\"true\">\n                    <option *ngIf=\"placeholder\" value=\"\">{{placeholder}}</option>\n                    <option *ngIf=\"selectedOption\" [value]=\"selectedOption.value\" [selected]=\"true\">{{selectedOption.label}}</option>\n                </select>\n            </div>\n            <label [ngClass]=\"{'ui-dropdown-label ui-inputtext ui-corner-all':true,'ui-dropdown-label-empty':(label == null || label.length === 0)}\" *ngIf=\"!editable && (label != null)\">\n                <ng-container *ngIf=\"!selectedItemTemplate\">{{label||'empty'}}</ng-container>\n                <ng-container *ngTemplateOutlet=\"selectedItemTemplate; context: {$implicit: selectedOption}\"></ng-container>\n            </label>\n            <label [ngClass]=\"{'ui-dropdown-label ui-inputtext ui-corner-all ui-placeholder':true,'ui-dropdown-label-empty': (placeholder == null || placeholder.length === 0)}\" *ngIf=\"!editable && (label == null)\">{{placeholder||'empty'}}</label>\n            <input #editableInput type=\"text\" [attr.aria-label]=\"selectedOption ? selectedOption.label : ' '\" class=\"ui-dropdown-label ui-inputtext ui-corner-all\" *ngIf=\"editable\" [disabled]=\"disabled\" [attr.placeholder]=\"placeholder\"\n                        (click)=\"onEditableInputClick($event)\" (input)=\"onEditableInputChange($event)\" (focus)=\"onEditableInputFocus($event)\" (blur)=\"onInputBlur($event)\">\n            <i class=\"ui-dropdown-clear-icon pi pi-times\" (click)=\"clear($event)\" *ngIf=\"value != null && showClear && !disabled\"></i>\n            <div class=\"ui-dropdown-trigger ui-state-default ui-corner-right\">\n                <span class=\"ui-dropdown-trigger-icon ui-clickable\" [ngClass]=\"dropdownIcon\"></span>\n            </div>\n            <div *ngIf=\"overlayVisible\" [ngClass]=\"'ui-dropdown-panel  ui-widget ui-widget-content ui-corner-all ui-shadow'\" [@overlayAnimation]=\"{value: 'visible', params: {showTransitionParams: showTransitionOptions, hideTransitionParams: hideTransitionOptions}}\" (@overlayAnimation.start)=\"onOverlayAnimationStart($event)\" [ngStyle]=\"panelStyle\" [class]=\"panelStyleClass\">\n                <div *ngIf=\"filter\" class=\"ui-dropdown-filter-container\" (click)=\"$event.stopPropagation()\">\n                    <input #filter type=\"text\" autocomplete=\"off\" [value]=\"filterValue||''\" class=\"ui-dropdown-filter ui-inputtext ui-widget ui-state-default ui-corner-all\" [attr.placeholder]=\"filterPlaceholder\"\n                    (keydown.enter)=\"$event.preventDefault()\" (keydown)=\"onKeydown($event, false)\" (input)=\"onFilter($event)\" [attr.aria-label]=\"ariaFilterLabel\">\n                    <span class=\"ui-dropdown-filter-icon pi pi-search\"></span>\n                </div>\n                <div class=\"ui-dropdown-items-wrapper\" [style.max-height]=\"virtualScroll ? 'auto' : (scrollHeight||'auto')\">\n                    <ul class=\"ui-dropdown-items ui-dropdown-list ui-widget-content ui-widget ui-corner-all ui-helper-reset\" role=\"listbox\">\n                        <ng-container *ngIf=\"group\">\n                            <ng-template ngFor let-optgroup [ngForOf]=\"optionsToDisplay\">\n                                <li class=\"ui-dropdown-item-group\">\n                                    <span *ngIf=\"!groupTemplate\">{{optgroup.label||'empty'}}</span>\n                                    <ng-container *ngTemplateOutlet=\"groupTemplate; context: {$implicit: optgroup}\"></ng-container>\n                                </li>\n                                <ng-container *ngTemplateOutlet=\"itemslist; context: {$implicit: optgroup.items, selectedOption: selectedOption}\"></ng-container>\n                            </ng-template>\n                        </ng-container>\n                        <ng-container *ngIf=\"!group\">\n                            <ng-container *ngTemplateOutlet=\"itemslist; context: {$implicit: optionsToDisplay, selectedOption: selectedOption}\"></ng-container>\n                        </ng-container>\n                        <ng-template #itemslist let-options let-selectedOption=\"selectedOption\">\n\n                            <ng-container *ngIf=\"!virtualScroll; else virtualScrollList\">\n                                <ng-template ngFor let-option let-i=\"index\" [ngForOf]=\"options\">\n                                    <p-dropdownItem [option]=\"option\" [selected]=\"selectedOption == option\" \n                                                    (onClick)=\"onItemClick($event,i)\"\n                                                    [template]=\"itemTemplate\"></p-dropdownItem>\n                                </ng-template>\n                            </ng-container>\n                            <ng-template #virtualScrollList>\n                                <cdk-virtual-scroll-viewport (scrolledIndexChange)=\"scrollToSelectedVirtualScrollElement($event)\" #viewport [ngStyle]=\"{'height': scrollHeight}\" [itemSize]=\"itemSize\" *ngIf=\"virtualScroll && optionsToDisplay && optionsToDisplay.length\">\n                                    <ng-container *cdkVirtualFor=\"let option of options; let i = index; let c = count; let f = first; let l = last; let e = even; let o = odd\">         \n                                        <p-dropdownItem [option]=\"option\" [selected]=\"selectedOption == option\"\n                                                                   (onClick)=\"onItemClick($event,i)\"\n                                                                   [template]=\"itemTemplate\"></p-dropdownItem>\n                                    </ng-container>\n                                </cdk-virtual-scroll-viewport>\n                            </ng-template>\n                        </ng-template>\n                        <li *ngIf=\"filter && optionsToDisplay && optionsToDisplay.length === 0\" class=\"ui-dropdown-empty-message\">{{emptyFilterMessage}}</li>\n                    </ul>\n                </div>\n            </div>\n        </div>\n    ",
                    animations: [
                        trigger('overlayAnimation', [
                            state('void', style({
                                transform: 'translateY(5%)',
                                opacity: 0
                            })),
                            state('visible', style({
                                transform: 'translateY(0)',
                                opacity: 1
                            })),
                            transition('void => visible', animate('{{showTransitionParams}}')),
                            transition('visible => void', animate('{{hideTransitionParams}}'))
                        ])
                    ],
                    host: {
                        '[class.ui-inputwrapper-filled]': 'filled',
                        '[class.ui-inputwrapper-focus]': 'focused'
                    },
                    providers: [DROPDOWN_VALUE_ACCESSOR]
                }] }
    ];
    /** @nocollapse */
    Dropdown.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: ChangeDetectorRef },
        { type: NgZone }
    ]; };
    Dropdown.propDecorators = {
        scrollHeight: [{ type: Input }],
        filter: [{ type: Input }],
        name: [{ type: Input }],
        style: [{ type: Input }],
        panelStyle: [{ type: Input }],
        styleClass: [{ type: Input }],
        panelStyleClass: [{ type: Input }],
        readonly: [{ type: Input }],
        required: [{ type: Input }],
        editable: [{ type: Input }],
        appendTo: [{ type: Input }],
        tabindex: [{ type: Input }],
        placeholder: [{ type: Input }],
        filterPlaceholder: [{ type: Input }],
        inputId: [{ type: Input }],
        selectId: [{ type: Input }],
        dataKey: [{ type: Input }],
        filterBy: [{ type: Input }],
        autofocus: [{ type: Input }],
        resetFilterOnHide: [{ type: Input }],
        dropdownIcon: [{ type: Input }],
        optionLabel: [{ type: Input }],
        autoDisplayFirst: [{ type: Input }],
        group: [{ type: Input }],
        showClear: [{ type: Input }],
        emptyFilterMessage: [{ type: Input }],
        virtualScroll: [{ type: Input }],
        itemSize: [{ type: Input }],
        autoZIndex: [{ type: Input }],
        baseZIndex: [{ type: Input }],
        showTransitionOptions: [{ type: Input }],
        hideTransitionOptions: [{ type: Input }],
        ariaFilterLabel: [{ type: Input }],
        onChange: [{ type: Output }],
        onFocus: [{ type: Output }],
        onBlur: [{ type: Output }],
        onClick: [{ type: Output }],
        onShow: [{ type: Output }],
        onHide: [{ type: Output }],
        containerViewChild: [{ type: ViewChild, args: ['container', { static: false },] }],
        filterViewChild: [{ type: ViewChild, args: ['filter', { static: false },] }],
        focusViewChild: [{ type: ViewChild, args: ['in', { static: false },] }],
        viewPort: [{ type: ViewChild, args: [CdkVirtualScrollViewport, { static: false },] }],
        editableInputViewChild: [{ type: ViewChild, args: ['editableInput', { static: false },] }],
        templates: [{ type: ContentChildren, args: [PrimeTemplate,] }],
        autoWidth: [{ type: Input }],
        disabled: [{ type: Input }],
        options: [{ type: Input }]
    };
    return Dropdown;
}());
export { Dropdown };
if (false) {
    /** @type {?} */
    Dropdown.prototype.scrollHeight;
    /** @type {?} */
    Dropdown.prototype.filter;
    /** @type {?} */
    Dropdown.prototype.name;
    /** @type {?} */
    Dropdown.prototype.style;
    /** @type {?} */
    Dropdown.prototype.panelStyle;
    /** @type {?} */
    Dropdown.prototype.styleClass;
    /** @type {?} */
    Dropdown.prototype.panelStyleClass;
    /** @type {?} */
    Dropdown.prototype.readonly;
    /** @type {?} */
    Dropdown.prototype.required;
    /** @type {?} */
    Dropdown.prototype.editable;
    /** @type {?} */
    Dropdown.prototype.appendTo;
    /** @type {?} */
    Dropdown.prototype.tabindex;
    /** @type {?} */
    Dropdown.prototype.placeholder;
    /** @type {?} */
    Dropdown.prototype.filterPlaceholder;
    /** @type {?} */
    Dropdown.prototype.inputId;
    /** @type {?} */
    Dropdown.prototype.selectId;
    /** @type {?} */
    Dropdown.prototype.dataKey;
    /** @type {?} */
    Dropdown.prototype.filterBy;
    /** @type {?} */
    Dropdown.prototype.autofocus;
    /** @type {?} */
    Dropdown.prototype.resetFilterOnHide;
    /** @type {?} */
    Dropdown.prototype.dropdownIcon;
    /** @type {?} */
    Dropdown.prototype.optionLabel;
    /** @type {?} */
    Dropdown.prototype.autoDisplayFirst;
    /** @type {?} */
    Dropdown.prototype.group;
    /** @type {?} */
    Dropdown.prototype.showClear;
    /** @type {?} */
    Dropdown.prototype.emptyFilterMessage;
    /** @type {?} */
    Dropdown.prototype.virtualScroll;
    /** @type {?} */
    Dropdown.prototype.itemSize;
    /** @type {?} */
    Dropdown.prototype.autoZIndex;
    /** @type {?} */
    Dropdown.prototype.baseZIndex;
    /** @type {?} */
    Dropdown.prototype.showTransitionOptions;
    /** @type {?} */
    Dropdown.prototype.hideTransitionOptions;
    /** @type {?} */
    Dropdown.prototype.ariaFilterLabel;
    /** @type {?} */
    Dropdown.prototype.onChange;
    /** @type {?} */
    Dropdown.prototype.onFocus;
    /** @type {?} */
    Dropdown.prototype.onBlur;
    /** @type {?} */
    Dropdown.prototype.onClick;
    /** @type {?} */
    Dropdown.prototype.onShow;
    /** @type {?} */
    Dropdown.prototype.onHide;
    /** @type {?} */
    Dropdown.prototype.containerViewChild;
    /** @type {?} */
    Dropdown.prototype.filterViewChild;
    /** @type {?} */
    Dropdown.prototype.focusViewChild;
    /** @type {?} */
    Dropdown.prototype.viewPort;
    /** @type {?} */
    Dropdown.prototype.editableInputViewChild;
    /** @type {?} */
    Dropdown.prototype.templates;
    /**
     * @type {?}
     * @private
     */
    Dropdown.prototype._autoWidth;
    /**
     * @type {?}
     * @private
     */
    Dropdown.prototype.virtualAutoScrolled;
    /**
     * @type {?}
     * @private
     */
    Dropdown.prototype.virtualScrollSelectedIndex;
    /**
     * @type {?}
     * @private
     */
    Dropdown.prototype._disabled;
    /** @type {?} */
    Dropdown.prototype.overlay;
    /** @type {?} */
    Dropdown.prototype.itemsWrapper;
    /** @type {?} */
    Dropdown.prototype.itemTemplate;
    /** @type {?} */
    Dropdown.prototype.groupTemplate;
    /** @type {?} */
    Dropdown.prototype.selectedItemTemplate;
    /** @type {?} */
    Dropdown.prototype.selectedOption;
    /** @type {?} */
    Dropdown.prototype._options;
    /** @type {?} */
    Dropdown.prototype.value;
    /** @type {?} */
    Dropdown.prototype.onModelChange;
    /** @type {?} */
    Dropdown.prototype.onModelTouched;
    /** @type {?} */
    Dropdown.prototype.optionsToDisplay;
    /** @type {?} */
    Dropdown.prototype.hover;
    /** @type {?} */
    Dropdown.prototype.focused;
    /** @type {?} */
    Dropdown.prototype.filled;
    /** @type {?} */
    Dropdown.prototype.overlayVisible;
    /** @type {?} */
    Dropdown.prototype.documentClickListener;
    /** @type {?} */
    Dropdown.prototype.optionsChanged;
    /** @type {?} */
    Dropdown.prototype.panel;
    /** @type {?} */
    Dropdown.prototype.dimensionsUpdated;
    /** @type {?} */
    Dropdown.prototype.selfClick;
    /** @type {?} */
    Dropdown.prototype.itemClick;
    /** @type {?} */
    Dropdown.prototype.clearClick;
    /** @type {?} */
    Dropdown.prototype.hoveredItem;
    /** @type {?} */
    Dropdown.prototype.selectedOptionUpdated;
    /** @type {?} */
    Dropdown.prototype.filterValue;
    /** @type {?} */
    Dropdown.prototype.searchValue;
    /** @type {?} */
    Dropdown.prototype.searchIndex;
    /** @type {?} */
    Dropdown.prototype.searchTimeout;
    /** @type {?} */
    Dropdown.prototype.previousSearchChar;
    /** @type {?} */
    Dropdown.prototype.currentSearchChar;
    /** @type {?} */
    Dropdown.prototype.documentResizeListener;
    /** @type {?} */
    Dropdown.prototype.el;
    /** @type {?} */
    Dropdown.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    Dropdown.prototype.cd;
    /** @type {?} */
    Dropdown.prototype.zone;
    /* Skipping unhandled member: ;*/
}
var DropdownModule = /** @class */ (function () {
    function DropdownModule() {
    }
    DropdownModule.decorators = [
        { type: NgModule, args: [{
                    imports: [CommonModule, SharedModule, ScrollingModule],
                    exports: [Dropdown, SharedModule, ScrollingModule],
                    declarations: [Dropdown, DropdownItem]
                },] }
    ];
    return DropdownModule;
}());
export { DropdownModule };
