// @bit/primefaces.primeng.internal.common.shared/dist/internal/common/shared/bundles/bit-internal.common.shared.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@bit/primefaces.primeng.internal.common.shared', ['exports', '@angular/core', '@angular/common'], factory) :
    (global = global || self, factory((global.bit = global.bit || {}, global.bit.internal = global.bit.internal || {}, global.bit.internal.common = global.bit.internal.common || {}, global.bit.internal.common.shared = {}), global.ng.core, global.ng.common));
}(this, function (exports, core, common) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var Header = /** @class */ (function () {
        function Header() {
        }
        Header.decorators = [
            { type: core.Component, args: [{
                        selector: 'p-header',
                        template: '<ng-content></ng-content>'
                    }] }
        ];
        return Header;
    }());
    var Footer = /** @class */ (function () {
        function Footer() {
        }
        Footer.decorators = [
            { type: core.Component, args: [{
                        selector: 'p-footer',
                        template: '<ng-content></ng-content>'
                    }] }
        ];
        return Footer;
    }());
    var PrimeTemplate = /** @class */ (function () {
        function PrimeTemplate(template) {
            this.template = template;
        }
        /**
         * @return {?}
         */
        PrimeTemplate.prototype.getType = /**
         * @return {?}
         */
        function () {
            return this.name;
        };
        PrimeTemplate.decorators = [
            { type: core.Directive, args: [{
                        selector: '[pTemplate]',
                        host: {}
                    },] }
        ];
        /** @nocollapse */
        PrimeTemplate.ctorParameters = function () { return [
            { type: core.TemplateRef }
        ]; };
        PrimeTemplate.propDecorators = {
            type: [{ type: core.Input }],
            name: [{ type: core.Input, args: ['pTemplate',] }]
        };
        return PrimeTemplate;
    }());
    /* Deprecated */
    var Column = /** @class */ (function () {
        function Column() {
            this.filterType = 'text';
            this.exportable = true;
            this.resizable = true;
            this.sortFunction = new core.EventEmitter();
        }
        /**
         * @return {?}
         */
        Column.prototype.ngAfterContentInit = /**
         * @return {?}
         */
        function () {
            var _this = this;
            this.templates.forEach((/**
             * @param {?} item
             * @return {?}
             */
            function (item) {
                switch (item.getType()) {
                    case 'header':
                        _this.headerTemplate = item.template;
                        break;
                    case 'body':
                        _this.bodyTemplate = item.template;
                        break;
                    case 'footer':
                        _this.footerTemplate = item.template;
                        break;
                    case 'filter':
                        _this.filterTemplate = item.template;
                        break;
                    case 'editor':
                        _this.editorTemplate = item.template;
                        break;
                    default:
                        _this.bodyTemplate = item.template;
                        break;
                }
            }));
        };
        Column.decorators = [
            { type: core.Component, args: [{
                        selector: 'p-column',
                        template: ''
                    }] }
        ];
        Column.propDecorators = {
            field: [{ type: core.Input }],
            colId: [{ type: core.Input }],
            sortField: [{ type: core.Input }],
            filterField: [{ type: core.Input }],
            header: [{ type: core.Input }],
            footer: [{ type: core.Input }],
            sortable: [{ type: core.Input }],
            editable: [{ type: core.Input }],
            filter: [{ type: core.Input }],
            filterMatchMode: [{ type: core.Input }],
            filterType: [{ type: core.Input }],
            excludeGlobalFilter: [{ type: core.Input }],
            rowspan: [{ type: core.Input }],
            colspan: [{ type: core.Input }],
            scope: [{ type: core.Input }],
            style: [{ type: core.Input }],
            styleClass: [{ type: core.Input }],
            exportable: [{ type: core.Input }],
            headerStyle: [{ type: core.Input }],
            headerStyleClass: [{ type: core.Input }],
            bodyStyle: [{ type: core.Input }],
            bodyStyleClass: [{ type: core.Input }],
            footerStyle: [{ type: core.Input }],
            footerStyleClass: [{ type: core.Input }],
            hidden: [{ type: core.Input }],
            expander: [{ type: core.Input }],
            selectionMode: [{ type: core.Input }],
            filterPlaceholder: [{ type: core.Input }],
            filterMaxlength: [{ type: core.Input }],
            frozen: [{ type: core.Input }],
            resizable: [{ type: core.Input }],
            sortFunction: [{ type: core.Output }],
            templates: [{ type: core.ContentChildren, args: [PrimeTemplate,] }],
            template: [{ type: core.ContentChild, args: [core.TemplateRef, { static: false },] }]
        };
        return Column;
    }());
    /* Deprecated */
    var Row = /** @class */ (function () {
        function Row() {
        }
        Row.decorators = [
            { type: core.Component, args: [{
                        selector: 'p-row',
                        template: ""
                    }] }
        ];
        Row.propDecorators = {
            columns: [{ type: core.ContentChildren, args: [Column,] }]
        };
        return Row;
    }());
    /* Deprecated */
    var HeaderColumnGroup = /** @class */ (function () {
        function HeaderColumnGroup() {
        }
        HeaderColumnGroup.decorators = [
            { type: core.Component, args: [{
                        selector: 'p-headerColumnGroup',
                        template: ""
                    }] }
        ];
        HeaderColumnGroup.propDecorators = {
            frozen: [{ type: core.Input }],
            rows: [{ type: core.ContentChildren, args: [Row,] }]
        };
        return HeaderColumnGroup;
    }());
    /* Deprecated */
    var FooterColumnGroup = /** @class */ (function () {
        function FooterColumnGroup() {
        }
        FooterColumnGroup.decorators = [
            { type: core.Component, args: [{
                        selector: 'p-footerColumnGroup',
                        template: ""
                    }] }
        ];
        FooterColumnGroup.propDecorators = {
            frozen: [{ type: core.Input }],
            rows: [{ type: core.ContentChildren, args: [Row,] }]
        };
        return FooterColumnGroup;
    }());
    var SharedModule = /** @class */ (function () {
        function SharedModule() {
        }
        SharedModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [common.CommonModule],
                        exports: [Header, Footer, Column, PrimeTemplate, Row, HeaderColumnGroup, FooterColumnGroup],
                        declarations: [Header, Footer, Column, PrimeTemplate, Row, HeaderColumnGroup, FooterColumnGroup]
                    },] }
        ];
        return SharedModule;
    }());

    exports.Column = Column;
    exports.Footer = Footer;
    exports.FooterColumnGroup = FooterColumnGroup;
    exports.Header = Header;
    exports.HeaderColumnGroup = HeaderColumnGroup;
    exports.PrimeTemplate = PrimeTemplate;
    exports.Row = Row;
    exports.SharedModule = SharedModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
